### DISCUSSION TITLE
Purpose of Folders (Videos/Uploads/Library) (Uploading all my media to new instance of immich)

### DISCUSSION THREAD

**Discussion by coltenkrauter** [immich-app/immich#14131] [Q&A] [answered]
Hey there, I have been running with TrueNas Scale Immich App for many months. It has been great, _but_ with the recent changes on TrueNas Scale to support Docker, I have decided to switch over to **Docker Compose**. So, temporarily, I will have 2 instances of immich running, while I find a way to fill up the new instance with all the photos/videos from the old instance.

Since there isn't a straightforward way to copy my photos/videos from my old Immich to my new immich _(meaning that I can't just mount the library... I think that this is designed so that the database must have reference of everything, and I don't think there is a scan option where the database will scan the folder structure and load in all the media)._

Anyway, so I figured the next best option would be to use the `immich-cli` to upload everything from my old instance to my new instance. 

So, I installed `immich-cli` such that it had access to the `Videos`, `Uploads`, and `Library` directories of my old instance... And then I logged the cli into my new instance and ran the following command, 

```sh
# current dir container Videos, Library, and Uploads directories.
sudo immich upload --recursive ./ -c 2
```

For more context, this is what the TrueNas Scale Immich setup looks like, 

![Screenshot 2024-11-13 at 4 57 00 PM](https://github.com/user-attachments/assets/10e6c6ca-dcd7-4b93-9801-61e9e148a71e)

---

## Why I am confused... the discrepancy.

Now, I was under the impression that the `Server Stats` reflected all that was within the `Videos`, `Uploads`, and `Library` directories.

Now, I ran `df -h ./*` and the used size of the directories are as follows, 
```
Library: 2.2 TB used
Uploads: 3.2 TB used
Videos: 482 GB used
```

And here is the server stats,

### OLD INSTANCE
![Screenshot 2024-11-13 at 5 27 23 PM](https://github.com/user-attachments/assets/7888a994-fc93-4924-9be0-3a05d9e8ac84)

### NEW INSTANCE
![Screenshot 2024-11-13 at 5 29 49 PM](https://github.com/user-attachments/assets/7f34d767-9b23-4f0f-937c-0211206e827c)

**Somehow, there are 10,000 more videos**... and 1000 more photos. 

## My questions

1. Why are there 10k more videos on the new instance when I simply uploaded the data from the old instance?
2. What is the purpose of the Videos/Uploads/Library folders?

Thanks for reading through this.

**Comment by alextran1502** [comment]
I am not sure what are store in the new Videos section on TrueNAS, my guess is that it includes transcoded video as well, so when you run the upload script, it also upload the transcoded video files as the new original files

**Reply by coltenkrauter** [reply]
Oh, this could be the case. When we transcode videos, are the originals preserved? 
If I knew the answer to this, then I might try to re-upload but omit the videos directory.

**Reply by alextran1502** [reply]
> When we transcode videos, are the originals preserved?

Yes, our design philosophy is that we don't touch the original files

**Reply by coltenkrauter** [reply]
Love that. Good choice.

**Comment by alextran1502** [comment]
> , and I don't think there is a scan option where the database will scan the folder structure and load in all the media

There is the external library feature https://immich.app/docs/features/libraries where you can mount existing library and have Immich scans it

**Reply by coltenkrauter** [reply]
Ah, that is a good suggestion. Are there any reasons why an internal library is better than external?

**Reply by alextran1502** [reply]
It depends on your use case. With the default library, newly uploaded files are stored there, while the external library uses a different mechanism, and the files are often managed from the file system.

**Reply by mertalev** [reply]
The advantages of an internal library are storage template migration (meaning that it can automatically organize uploaded photos) and checksum deduplication (meaning it will reject a second asset with the same checksum).

External libraries have the benefits of file watching, less storage overhead (if you still need to keep the files as they are, separate from Immich's file structure), and faster importing as they can scan the files directly without the ceremony of an upload client, etc.

**Comment by coltenkrauter** [comment]
I am still curious what the purpose of the uploads directory is. Do uploads stay in the uploads directory forever?

So, if I am interested in using the immich CLI to upload all my photos and videos to my new instance, if I just target the uploads directory as the source, will that upload all of my media to my new instance of immich?

What is the library directory holding? Duplicates of the uploads folder?

**Reply by alextran1502** [reply, marked as answer]
When you upload an asset, it first arrives at the `upload` directory, then if you have the `Storage Template` feature enabled, it will extract the needed metadata and move the file to the specified folder structure in the `library` folder

**Reply by coltenkrauter** [reply]
Ah, makes sense. So, between the uploads directory and the library directory, I should be able to find all of my media. 

Is there anything else stored in the library directory?

**Reply by alextran1502** [reply]
If you have the Storage template feature enabled, then it stores your original files in the `Library`, and you can find some interrupted upload files in the `upload` folder. If the storage template is not used then all the upload files can be found in the `upload` folder

**Comment by coltenkrauter** [comment]
Okay, another follow-up here... I'll see if anyone can help in this thread before I open up a new discussion or issue.

---

I decided to start over and use the immich cli to upload only the `uploads` and `library` directories from my old immich instance to my new immich instance. 

Now, that has worked better... _Except_, it seems (as mentioned earlier in this thread) that there were a large number of `interrupted upload files` in the `uploads` directory. So, those all got uploaded along with all of the other media. So, now, I have all of my original photos and videos with partial images and videos that will not play or display properly, even when I click into them or download them. 

---
I have 2 questions...

**1. Is there anyway to bulk delete all of these corrupted _partial uploads_?**

**2. Are there any plans to add some ability to clean these up in the future of immich, because I don't want partial uploads sticking around wasting space on my HDD?**

**Reply by alextran1502** [reply]
1. You can enabled storage template on your old instance, then run the storage template migration job, it will move all good files to the `library` folder. Then you can just ignore the `upload` folder

2. Yes

**Reply by coltenkrauter** [reply]
Hey, thanks a lot for the quick response. 

1. Good thinking. I'll do as you said.

2. Awesome. Love this product.

**Reply by coltenkrauter** [reply]
How does the storage migration job work? Is it one job to move all the assets? Or one job per file type? Or something else?

I started the job, and it says 1 in progress, 6 waiting. I'm curious how it determines how many jobs to queue up.

![Screenshot 2024-11-16 at 10 54 36 AM](https://github.com/user-attachments/assets/c9f57a37-1882-4b80-865c-b763a6406f55)

**Reply by mertalev** [reply]
It's one job for all assets. I'm actually not sure why there are 6 waiting.

**Reply by coltenkrauter** [reply]
Hmmm, that is odd. I can tell you that I have 3 or 4 TB of data, so I wonder if that has something to do with it. 🤷🏻

**Comment by coltenkrauter** [comment]
Description of folders [here](https://github.com/immich-app/immich/discussions/17493#discussioncomment-12783584)
